File: go/0001_two_sum.go
package main

// time: O(n)
// space: O(n)

func twoSum(numbers []int, target int) []int {
	hash := map[int]int{}

	for i, n := range numbers {
		diff := target - n
		idx, contains := hash[diff]
		if contains {
			return []int{idx, i}
		}
		hash[n] = i
	}
	return nil
}

File: typescript/0001-two-sum.ts
function twoSum(nums: number[], target: number): number[] {
  let map: { [key: number]: number } = {}

  for (let i = 0; i < nums.length; i++){
    const diff = target - nums[i]
    if(diff in map){
      return [map[diff], i]
    }

    map[nums[i]] = i
  }

  return []
}
